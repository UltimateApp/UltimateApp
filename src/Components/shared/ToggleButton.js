import React from 'react';
import { StyleSheet } from 'react-native';
import { ToggleButton as PaperToggleButton } from 'react-native-paper';

import theme from '../../styles/theme.style';

const ToggleButton = ({ value, onValueChange, possibleValues, icons }) => {
  const renderButton = (possibleValue, index) => {
    const pressed = possibleValue === value;
    const positionalStyle =
      index === 0 ? styles.first : index === possibleValues.length - 1 ? styles.last : styles.middle;
    return (
      <PaperToggleButton
        key={index}
        color={pressed ? theme.MAIN_COLOR : undefined}
        style={[styles.button, positionalStyle, pressed ? styles.pressedToggleButton : undefined]}
        icon={icons[index]}
        value={possibleValue}
        disabled={pressed}
        testID={`toggle-${possibleValue}`}
      />
    );
  };

  return (
    <PaperToggleButton.Row style={styles.row} value={value} onValueChange={onValueChange}>
      {possibleValues.map(renderButton)}
    </PaperToggleButton.Row>
  );
};

const styles = StyleSheet.create({
  row: {
    marginTop: 5,
    justifyContent: 'center',
  },
  button: {
    borderColor: theme.COLOR_SECONDARY,
    borderWidth: StyleSheet.hairlineWidth,
  },
  pressedToggleButton: {
    backgroundColor: theme.COLOR_SECONDARY_LIGHT,
    opacity: 0.8,
  },
  first: {
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0,
  },
  middle: {
    borderRadius: 0,
    borderLeftWidth: 0,
  },
  last: {
    borderLeftWidth: 0,
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
  },
});

export default ToggleButton;
