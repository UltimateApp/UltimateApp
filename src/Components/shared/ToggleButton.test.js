import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';

import ToggleButton from './ToggleButton';

describe('<ToggleButton />', () => {
  const onValueChange = jest.fn();
  const value = 'animation';
  const icons = ['clipboard-outline', 'video'];
  const possibleValues = ['animation', 'video'];

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ToggleButton value={value} onValueChange={onValueChange} icons={icons} possibleValues={possibleValues} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('delegates onValueChange prop', async () => {
    const { getByTestId } = render(
      <ToggleButton value={value} onValueChange={onValueChange} icons={icons} possibleValues={possibleValues} />,
    );

    await fireEvent.press(getByTestId('toggle-video'));

    expect(onValueChange).toHaveBeenCalledWith('video');
  });
});
