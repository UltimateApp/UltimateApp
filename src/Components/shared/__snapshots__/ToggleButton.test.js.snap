// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButton /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      Object {
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "margin": 6,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 18,
            "height": 36,
            "width": 36,
          },
          Object {
            "opacity": 0.32,
          },
          Array [
            Object {
              "height": 42,
              "margin": 0,
              "width": 42,
            },
            Object {
              "backgroundColor": "rgba(0, 0, 0, .08)",
              "borderColor": "rgba(0, 0, 0, 0.29)",
              "borderRadius": 4,
            },
            Array [
              Object {
                "borderWidth": 0.5,
              },
              Object {
                "borderBottomRightRadius": 0,
                "borderTopRightRadius": 0,
              },
              Object {
                "backgroundColor": "#f1f1f1",
                "opacity": 0.8,
              },
            ],
          ],
        ],
      ]
    }
    testID="toggle-animation"
  >
    <View>
      <Text />
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "margin": 6,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 18,
            "height": 36,
            "width": 36,
          },
          false,
          Array [
            Object {
              "height": 42,
              "margin": 0,
              "width": 42,
            },
            Object {
              "backgroundColor": "transparent",
              "borderColor": "rgba(0, 0, 0, 0.29)",
              "borderRadius": 4,
            },
            Array [
              Object {
                "borderWidth": 0.5,
              },
              Object {
                "borderBottomLeftRadius": 0,
                "borderLeftWidth": 0,
                "borderTopLeftRadius": 0,
              },
              undefined,
            ],
          ],
        ],
      ]
    }
    testID="toggle-video"
  >
    <View>
      <Text />
    </View>
  </View>
</View>
`;
